File: src/2023-04-13-custom_rpc_options/go/main.go
package main

import (
	"fmt"

	pb "github.com/Clement-Jean/test/proto"
	"google.golang.org/protobuf/proto"
	"google.golang.org/protobuf/reflect/protoreflect"
	"google.golang.org/protobuf/types/descriptorpb"
)

func getOptionValue[T string | int | bool](
	opts *descriptorpb.MethodOptions,
	id protoreflect.ExtensionType,
) *T {
	value, ok := proto.GetExtension(opts, id).(T)

	if ok {
		return &value
	}

	return nil
}

func getServiceMethod(
	sd protoreflect.ServiceDescriptor,
	fn func(md protoreflect.MethodDescriptor) bool,
) *protoreflect.MethodDescriptor {
	for i := 0; i < sd.Methods().Len(); i++ {
		md := sd.Methods().Get(i)

		if fn(md) {
			return &md
		}
	}

	return nil
}

func getMethodOptionValue[T string | int | bool](
	sd protoreflect.ServiceDescriptor,
	id protoreflect.ExtensionType,
) *T {
	var value *T = nil

	getServiceMethod(sd, func(md protoreflect.MethodDescriptor) bool {
		opts, ok := md.Options().(*descriptorpb.MethodOptions)

		if !ok {
			return false
		}

		if tmp := getOptionValue[T](opts, id); tmp != nil {
			value = tmp
			return true
		}

		return false
	})

	return value
}

func getMethodExtension[T string | int | bool](
	fd protoreflect.FileDescriptor,
	id protoreflect.ExtensionType,
) *T {
	for i := 0; i < fd.Services().Len(); i++ {
		sd := fd.Services().Get(i)

		if value := getMethodOptionValue[T](sd, id); value != nil {
			return value
		}
	}

	return nil
}

func main() {
	world := getMethodExtension[string](pb.File_proto_world_proto, pb.E_Hello)

	if world != nil {
		fmt.Println(*world)
	}
}

File: src/2023-04-13-custom_rpc_options/java/main.java
package com.clementjean.test;

import com.google.protobuf.Descriptors;
import com.google.protobuf.DescriptorProtos;
import com.google.protobuf.GeneratedMessage;
import java.util.Optional;
import java.util.function.Function;
import java.util.concurrent.atomic.AtomicReference;

class HelloWorldApp {
    private static <T> Optional<T> getOptionValue(
        DescriptorProtos.MethodOptions opts,
        GeneratedMessage.GeneratedExtension<DescriptorProtos.MethodOptions, ?> id
    ) {
        return opts.hasExtension(id) ?
              Optional.of((T)opts.getExtension(id)) :
              Optional.empty();
    }

    private static <T> Optional<Descriptors.MethodDescriptor> getServiceMethod(
        Descriptors.ServiceDescriptor sd,
        Function<Descriptors.MethodDescriptor, Boolean> predicate
    ) {
        for (int i = 0; i < sd.getMethods().size(); ++i) {
            Descriptors.MethodDescriptor method = sd.getMethods().get(i);

            if (predicate.apply(method))
                return Optional.of(method);
        }

        return Optional.empty();
    }

    private static <T> Optional<T> getMethodOptionValue(
        Descriptors.ServiceDescriptor sd,
        GeneratedMessage.GeneratedExtension<DescriptorProtos.MethodOptions, ?> id
    ) {
        AtomicReference<Optional<T>> value = new AtomicReference<>(Optional.empty());

        getServiceMethod(sd, md -> {
            DescriptorProtos.MethodOptions opts = md.getOptions();
            Optional<T> tmp = getOptionValue(opts, id);

            if (tmp.isPresent())
                value.set(tmp);

            return value.get().isPresent();
        });

		return value.get();
	}

    private static <T> Optional<T> getMethodExtension(
        Descriptors.FileDescriptor fd,
        GeneratedMessage.GeneratedExtension<DescriptorProtos.MethodOptions, ?> id
    ) {
        for (int i = 0; i < fd.getServices().size(); ++i) {
            Descriptors.ServiceDescriptor sd = fd.getServices().get(i);
            Optional<T> world = getMethodOptionValue(sd, Hello.hello);

            if (world.isPresent())
                return world;
        }

        return Optional.empty();
    }

    public static void main(String[] args) {
        Descriptors.FileDescriptor fd = World.getDescriptor();
        Optional<String> world = getMethodExtension(fd, Hello.hello);

        world.ifPresent(w -> System.out.println(w));
    }
}